lhotse/dataset/collation.py

```diff
@@ -1,11 +1,109 @@
-from typing import Iterable, Union
+from typing import Iterable, Union, Literal, Tuple, Callable
+from dataclasses import dataclass
 
 import numpy as np
 import torch
 from torch.nn import CrossEntropyLoss
 
 from lhotse import CutSet
 from lhotse.cut import MixedCut
+from lhotse.utils import LOG_EPSILON
+
+
+@dataclass
+class CollateTokens:
+
+    tokens: Iterable[str]
+    add_eos: bool = True
+    add_bos: bool = True
+    pad_symbol: str = "<pad>"
+    bos_symbol: str = "<bos>"
+    eos_symbol: str = "<eos>"
+    unk_symbol: str = "<unk>"
+
+    def __post_init__(self):
+        tokens_unique = (
+            [self.pad_symbol, self.unk_symbol]
+            + ([self.bos_symbol] if self.add_bos else [])
+            + ([self.eos_symbol] if self.add_eos else [])
+            + sorted(set(self.tokens))
+        )
+        self.token2int = {token: idx for idx, token in enumerate(tokens_unique)}
+
+    def __call__(self, cuts: CutSet) -> Tuple[torch.Tensor, torch.Tensor]:
+        token_sequences = [cut.supervisions[0].text for cut in cuts]
+        max_len = len(max(token_sequences, key=len))
+        seqs = [
+            ([self.bos_symbol] if self.add_bos else [])
+            + list(seq)
+            + ([self.eos_symbol] if self.add_eos else [])
+            + [self.pad_symbol] * (max_len - len(seq))
+            for seq in token_sequences
+        ]
+
+        tokens_batch = torch.from_numpy(np.array([[self.token2int[token] for token in seq] for seq in seqs], dtype=np.int))
+        tokens_lens = torch.tensor([len(seq) for seq in token_sequences], dtype=torch.int)
+        return tokens_batch, tokens_lens
+
+
+# This will go to some normalization utils - leaving it here for now for readability
+@dataclass
+class StandardizeFeatures:
+    norm_means: np.ndarray
+    norm_stds: np.ndarray
+
+    def __call__(self, features):
+        return (features - self.norm_means) / self.norm_stds
+
+    def inverse(self, features):
+        return features * self.norm_stds + self.norm_means
+
+
+@dataclass
+class RescaleFeatures:
+    feature_min: float
+    feature_max: float
+    new_min: float = 0.0
+    new_max: float = 1.0
+
+    def __post_init__(self):
+        self.old_range = self.feature_max - self.feature_min
+        self.new_range = self.new_max - self.new_min
+
+    def __call__(self, features):
+        return self.new_min + (features - self.feature_min) * self.new_range / self.old_range
+
+    def inverse(self, features):
+        return self.feature_min + (features - self.new_min) * self.old_range / self.new_range
+
+
+@dataclass
+class CollateFeatures:
+    pad_mode: Literal["constant", "reflect"] = "constant"
```

Comment:
Why do we need "reflect"?

Reply:
Reflect can be used for batch norm to have unbiased estimate for batch variance (which usually happens if zero padding or other constant is used.) I put this here to show what configurations could be used and left the implementation for this blank. We can completely purge it for now.